### ISSUE TITLE
Office.context.mailbox.item is null

### ISSUE THREAD

**Issue by ArjunSureshKumar92** [OfficeDev/office-js#1402]
I have created an outlook add-in for our customers. For only one customer when we retrieve the Office.context.mailbox.item object, we are getting a null value. This issue only occurs for this customer and others have no issue.

The issue is only with the outlook desktop client. OWA is working fine. Customer is using windows 7/windows 2019 with outlook version 2002.

`Office.onReady(function () {
Office.initialize = function () {}

       refresh: function () {
                try {
                    app.logtext += "\nRefreshing data";
                    if (!Office.initialize) {
                        app.logtext += "\nError initializing office..."
                    } else {
                        if (Office.context) {
                            app.logtext += "\nOffice.context true"
                        }
                        if (Office.context.mailbox) {
                            app.logtext += "\nOffice.context.mailbox true"
                        }
                        if (Office.context.mailbox.item) {
                            app.logtext += "\nOffice.context.mailbox.item true"
                        }
                        app.logtext += "\nOffice.context.mailbox.item.internetMessageId" + Office.context.mailbox.item.internetMessageId;
                        app.logtext += "\nOffice.context.mailbox.item.internetMessageId" + Office.context.mailbox.item.internetMessageId;
                        app.logtext += "\nOffice.context.mailbox.item.internetMessageId" + Office.context.mailbox.item.internetMessageId;
                        app.logtext += "\nOffice.context.mailbox.item.subject" + Office.context.mailbox.item.subject;
                        app.logtext += "\nOffice.context.mailbox.item.itemId" + Office.context.mailbox.item.itemId;
                    }
                    app.internetMsgId = Office.context.mailbox.item.internetMessageId;
                    app.subject = Office.context.mailbox.item.subject;
                    app.messageId = Office.context.mailbox.item.itemId;
                } catch (err) {
                    app.logtext += "\n JS error" + err;
                }
                app.settings_egCookie = Office.context.roamingSettings.get("Addin-Cookie");
                app.settings_userEmail = Office.context.roamingSettings.get("Addin-User-Email");
                app.settings_userName = Office.context.roamingSettings.get("Addin-User-Name");
                app.settings_userId = Office.context.roamingSettings.get("Addin-User-Id");
                app.settings_clientId = Office.context.roamingSettings.get("Addin-Client-Id");
                app.emailgisticsUser = Office.context.roamingSettings.get("Addin-Emailgistics-User");
                
            },

        }
});

$(document).ready(function () {
    Office.context.mailbox.addHandlerAsync(Office.EventType.ItemChanged, itemChanged);
    //UpdateTaskPaneUI(Office.context.mailbox);
    app.logtext += "\Document is ready and task pane is loaded....";
});
});`

// if (Office.context.mailbox.item)  this condition is failing and the data is null from my tests.

**Comment by exextoc** [comment]
Is the add-in pinned? If the add-in is pinned and there is an event handler for an ItemChanged event, the item can be null when you switch items or select an item header. Can you describe how your add-in is being used?

**Comment by ArjunSureshKumar92** [comment]
I tried the same without pinning the addin, the result is the same. Like I stated in the description is it working fine for me and most of my customers. We only have this issue with one customer. 
_If the add-in is pinned and there is an event handler for an ItemChanged event_
Correct me if I'm wrong, but even if we have this event handler, the value we get from the office object should not be null right? When I call the office.context.mailbox is it valid, it is only the office.context.mailbox.item that is null. 

Is there any security policy that will prevent the outlook add-in from accessing this info? If so, kindly provide a link. 
Thanks.

**Comment by exextoc** [comment]
Can you provide repro steps or a screen recording of how you are getting this error.

**Comment by ElizabethSamuel-MSFT** [comment]
@ArjunSureshKumar92 Can you provide repro steps or recording per exextoc's earlier request? Thx.

**Comment by ArjunSureshKumar92** [comment]
@ElizabethSamuel-MSFT  Sorry for the delay. Attaching screen shots for the steps.
Explanation:
Success Case:
Step 1(success_1.png):
Select the test addin
Step 2(success_2.png):
Press refresh button, and the message details is populated correctly. message id, message imi and subject
Step 3(success_3.png):
The log info is shown.
All the office object values are present so we return true.

Error case for only one customer:
Step1(error_1.png):
Select the test add-in
Step2(error_2.png):
Press the refresh button.
Step3(error_3.png):
The message details are not popuated.
Step4(error_4.png) and (error_5.png):
The log shows the error.
so the Office.context.mailbox.item value is null, so the data is missing.
![success_1](https://user-images.githubusercontent.com/35875176/95459274-6c558400-0941-11eb-8f65-f60673e9c91e.png)
![success_2](https://user-images.githubusercontent.com/35875176/95459276-6c558400-0941-11eb-8084-290bb3cdcea6.png)
![success_3](https://user-images.githubusercontent.com/35875176/95459277-6cee1a80-0941-11eb-9858-e5f9a79695ab.png)
![error_1](https://user-images.githubusercontent.com/35875176/95459279-6cee1a80-0941-11eb-8a81-aa970004e4cd.png)
![error_2](https://user-images.githubusercontent.com/35875176/95459282-6d86b100-0941-11eb-9bf1-b3896aaed355.png)
![error_3](https://user-images.githubusercontent.com/35875176/95459283-6d86b100-0941-11eb-8715-87a878dc35d4.png)
![error_4](https://user-images.githubusercontent.com/35875176/95459284-6d86b100-0941-11eb-8e25-fc194d41f48a.png)
![error_5](https://user-images.githubusercontent.com/35875176/95459286-6e1f4780-0941-11eb-81cf-a4f7565927f9.png)

**Comment by exextoc** [comment]
- Does your refresh do anything aside from just dump stuff out to the log? (from the code you pasted, it doesn't look like it, but I wanted to double check). 

- Instead of a Refresh button could you just do everything in the Initialize() function? Also just use OnReady() OR Initialize(). Not both. https://docs.microsoft.com/en-us/office/dev/add-ins/develop/initialize-add-in. It could be the fact that you are defining Initialize() inside OnReady is causing issues. 

In the past, we've only seen problems with this when Item is accessed outside Initialize/OnReady(). Your code looks okay, but the fact that you have both Initialize and OnReady(), seems somewhat suspicious. 

- Does this happen on EVERY email of this customer? 

- Does the customer have an up to date version of your manifest add-in? Their Add-in looks out of date (it doesn't have the pinning icon)

- Do other add-ins on this customers machine behave incorrectly? (notably if you use ScriptLab do the example scripts work?)

**Comment by ArjunSureshKumar92** [comment]
_Does your refresh do anything aside from just dump stuff out to the log? (from the code you pasted, it doesn't look like it, but I wanted to double check)._
No, this is a debugging addin we created to strip out everything other than getting the item from the office js object.

_Instead of a Refresh button could you just do everything in the Initialize() function? Also just use OnReady() OR Initialize(). Not both. https://docs.microsoft.com/en-us/office/dev/add-ins/develop/initialize-add-in. It could be the fact that you are defining Initialize() inside OnReady is causing issues._
I will make changes to the code to see if this fixes the issue. But since only one customer is experiencing the problem, I'm not that confident.

_Does this happen on EVERY email of this customer?_
Yes. We must have tried with at least 5 or 6 emails, and all had the same issue. However, something interesting we found was, with OWA, the same emails all worked fine. Which is why I still believe it might be a office desktop client security setting.

_Does the customer have an up to date version of your manifest add-in? Their Add-in looks out of date (it doesn't have the pinning icon)_
Yes, we checked the manifest and confirmed the version is the same. Like you said, the pinning icon is missing, which we noticed, but our best guess is it might be a windows version issue. We are using windows 10, but the customer is using windows 7.
Is there any known issue with windows 7 and office js?

_Do other add-ins on this customers machine behave incorrectly? (notably if you use ScriptLab do the example scripts work?)_
Unfortunately, we cant request this test with our customer. 

Note: As per your comments, I will update the code to remove the initialize function, and then connect with the customer and provide screenshots asap.

**Comment by ArjunSureshKumar92** [comment]
Code Update:
`Office.onReady(function () {

refresh: function () {
            try {
                app.logtext += "\nRefreshing data";
                if (!Office.initialize) {
                    app.logtext += "\nError initializing office..."
                } else {
                    if (Office.context) {
                        app.logtext += "\nOffice.context true"
                    }
                    if (Office.context.mailbox) {
                        app.logtext += "\nOffice.context.mailbox true"
                    }
                    if (Office.context.mailbox.item) {
                        app.logtext += "\nOffice.context.mailbox.item true"
                    }
                    app.logtext += "\nOffice.context.mailbox.item.internetMessageId" + Office.context.mailbox.item.internetMessageId;
                    app.logtext += "\nOffice.context.mailbox.item.internetMessageId" + Office.context.mailbox.item.internetMessageId;
                    app.logtext += "\nOffice.context.mailbox.item.internetMessageId" + Office.context.mailbox.item.internetMessageId;
                    app.logtext += "\nOffice.context.mailbox.item.subject" + Office.context.mailbox.item.subject;
                    app.logtext += "\nOffice.context.mailbox.item.itemId" + Office.context.mailbox.item.itemId;
                }
                app.internetMsgId = Office.context.mailbox.item.internetMessageId;
                app.subject = Office.context.mailbox.item.subject;
                app.messageId = Office.context.mailbox.item.itemId;
            } catch (err) {
                app.logtext += "\n JS error" + err;
            }
            app.settings_egCookie = Office.context.roamingSettings.get("Addin-Cookie");
            app.settings_userEmail = Office.context.roamingSettings.get("Addin-User-Email");
            app.settings_userName = Office.context.roamingSettings.get("Addin-User-Name");
            app.settings_userId = Office.context.roamingSettings.get("Addin-User-Id");
            app.settings_clientId = Office.context.roamingSettings.get("Addin-Client-Id");
            app.emailgisticsUser = Office.context.roamingSettings.get("Addin-Emailgistics-User");
            
        },

    }`
    

_Office.initialize = function () {} _ removed.

There was no change. The error is still the same.

**Comment by exextoc** [comment]
You mentioned the customer is using both Windows 7 and Windows Server 2019? Does the issue occur on both configurations? It could be unrelated, but that version of Outlook (Version 2002) [is supported on](https://www.microsoft.com/en-us/microsoft-365/microsoft-365-and-office-resources) Windows Server 2019, but not on Windows 7.

**Comment by ArjunSureshKumar92** [comment]
We had the issue with both, but recently all the tests were on Windows 7. 
_but that version of Outlook (Version 2002) is supported on Windows Server 2019, but not on Windows 7._
So if I understand correctly, this is a version issue? Can you provide me a link where this info is available. It will be easier for me to convince the customer to upgrade with that link available for them to check as well. Thanks.

**Comment by exextoc** [comment]
Windows 7 support has ended. More information can be found [here](https://support.microsoft.com/en-us/windows/windows-7-support-ended-on-january-14-2020-b75d4580-2cc7-895a-2c9c-1466d9a53962).

You can check if a version of Outlook is supported on Windows Server 2019 [here](https://www.microsoft.com/en-us/microsoft-365/microsoft-365-and-office-resources).

**Comment by ArjunSureshKumar92** [comment]
Got the customer to update to Windows 10, but still have the exact same issue.
Update info: 
Windows 10: 1909
Outlook version(o365) : 2002
MSO version: 16.0.12

We are close to losing the customer, any help will be much appreciated. Thanks

**Comment by ArjunSureshKumar92** [comment]
@ElizabethSamuel-MSFT 
Is it possible for IRM or other security features to prevent the add-in from getting the item info.

**Comment by ElizabethSamuel-MSFT** [comment]
@ArjunSureshKumar92 Thanks for providing this additional info.

@exextoc Can you take another look?

Thanks.

**Comment by exextoc** [comment]
Are you using IRM or signed messages? To activate add-ins on IRM make sure that you have the requirements documented [here](https://docs.microsoft.com/en-us/office/dev/add-ins/reference/objectmodel/preview-requirement-set/outlook-requirement-set-preview#add-in-activation-on-items-protected-by-information-rights-management-irm) and [here](https://docs.microsoft.com/en-us/office/dev/add-ins/outlook/outlook-add-ins-overview#mailbox-items-available-to-add-ins).

**Comment by exextoc** [comment]
With IRM, the behavior should be that the addin's would not open and addins should not be available.

I want to verify your repro here. does Office.onReady get called by the framework? (note that the framework / office.js+outlook.js should call Office.onReady). API's are not guaranteed to be ready. Office* APIs (including Office.context.mailbox.item) are only ready during or after Office.onReady/initialize is called). 

The other thing to check - have you checked the console log to see if there are errors there.

Another thing I notice, but probably not related to the error is that during $document.onReady, as noted above, the office.* api's aren't gaurenteed to be ready. addHandlerAsync should be called during Office.onReady/initialize.

**Comment by ArjunSureshKumar92** [comment]
_ does Office.onReady get called by the framework? _

The entire code is written inside office ready.
`Office.onReady(function () {
// the entire code...
});`

_The other thing to check - have you checked the console log to see if there are errors there._
Not sure what you mean, one of our customers is facing this issue in their outlook desktop version. Not sure how I can get the console logs. Can you point me to a example link.

Thanks.

**Comment by exextoc** [comment]
Got it on Office.onReady. I wasn't sure if the result was because of clicking on refresh, or if the result was filled in from the code in Office.onReady
In terms of console log, a debugger needs to be attached https://docs.microsoft.com/en-us/office/dev/add-ins/testing/debug-add-ins-using-f12-developer-tools-on-windows-10 To do this, you will need access to the customer's repro.

**Comment by ArjunSureshKumar92** [comment]
> _In terms of console log, a debugger needs to be attached https://docs.microsoft.com/en-us/office/dev/add-ins/testing/debug-add-ins-using-f12-developer-tools-on-windows-10 To do this, you will need access to the customer's repro._

Yes, I will try to get this info asap. Thanks.

**Comment by ArjunSureshKumar92** [comment]
We are currently working with Microsoft support. Thank you for  all your help. Closing the case for now.

**Comment by wsoccorsi-yw** [comment]
👋  @exextoc I used to experience something similar but no longer do, I'm assuming this was patched. Would you be able to point to documentation around this `ItemChanged` `null` `item `fix?

**Comment by ArjunSureshKumar92** [comment]
Our interaction with Office 365 support, [Case #:22924777]  has been closed with the following reason: "the issue needs to be fixed by developer team". From our tests, we've noticed for a tenant having 2 shared mailboxes, one has the issue while the other works fine. This rules out a tenant specific issue, or an underlying network issue. There is no difference in the way the shared mailboxes were created on our end, so we assume it has to be something internal that is failing.
Any support will be much appreciated.

**Comment by exextoc** [comment]
Could you share the console.log information as requested previously. Also, post the update was the user able to see the pin icon in their Outlook, given their manifest is the most recent one like you confirmed?

**Comment by exextoc** [comment]
Also, add-ins don't have any official support for the [Shared Mailboxes](https://docs.microsoft.com/en-us/microsoft-365/admin/email/create-a-shared-mailbox?view=o365-worldwide) feature in Outlook. Is that what you were referring to in your last response, or is that completely unrelated?

**Comment by mwickens** [comment]
Thanks for the response. Arjun is out of the office until Monday, but I'll try to get the info you've requested.

I can tell you that, yes, the add-in is designed to work with (and only with) shared mailboxes. As for them not being supported, is [this page](https://docs.microsoft.com/en-us/office/dev/add-ins/reference/manifest/supportssharedfolders) wrong?

**Comment by exextoc** [comment]
The page/documentation you referred to is about Shared Folders, not Shared Mailboxes. Those are 2 separate features. So it appears the issue you are currently having is only in a Shared Folders scenario?

**Comment by mwickens** [comment]
Our issue is definitely with shared mailboxes. Sorry for my confusion, and sorry if I linked to an irrelevant page, but we followed along on community calls and discussed shared mailbox support with several Microsoft people as it was being developed over a year ago. [This user voice page ](https://officespdev.uservoice.com/forums/224641-feature-requests-and-feedback/suggestions/11634318-support-for-shared-mailboxes) announced its release as part of the 1.8 requirement set. And, to be clear, it works fine except for this rare issue that crops up with random mailboxes.

**Comment by exextoc** [comment]
I'm checking with our User-Voice page admin. I will send an update once I hear back from the admin.

**Comment by exextoc** [comment]
I have informed our User-Voice page admin. It appears our documentation and User-Voice are conflicting. We are working to correct this issue.

**Comment by ArjunSureshKumar92** [comment]
@exextoc 
> Could you share the console.log information as requested previously. Also, post the update was the user able to see the pin icon in their Outlook, given their manifest is the most recent one like you confirmed?

We do not have the console.log available. Unfortunately, we were not able to install the debugger in the customers system. 

The user is able to see the pin icon in their outlook.

Is there any other info we can provide that would help?.

Thanks.

**Comment by vputsenko** [comment]
Hi @exextoc,

OfficeJS raises itemChanged event in next situations:

1. each time user changes selection to unsupported item 
2. user switches between compose and read mode - Outlook runs two sidebars in this case, one is hidden another is visible.
3. even simpler - on each item change OfficeJS raises itemChanged event passing null item following by ItemChanged event with selected item.  

You can even check your source code and you also have telemetry. Look at next code:
![image](https://user-images.githubusercontent.com/10497743/108846600-6ce87180-75e7-11eb-980f-d72d43f24bb4.png)

There is an explicit assignment to null. This is https://appsforoffice.microsoft.com/lib/1.1/hosted/outlook-win32-16.02.js  (line 3432 if you press "pretty print"). Obviously this is intentional behavior which is not documentation and doesn't have corresponding typing for Typescript. So i believe it is better just document and fix typings so TS code will fail to compile if necessary conditions are not added.

Please confirm that it  helped you to have repro at least.

@ElizabethSamuel-MSFT FYI.

**Comment by exextoc** [comment]
I think something may have gotten lost as we passed this issue between owners. There are also several users commenting in here on several different issues.

What you describe is intended behavior as you noted. (and we had mentioned before that Null Item while pinning needs to be accounted for). Null items should NOT be coming through when not pinned. (A previous post said that NULL items were coming through for EVERY item regardless of pinning)

Outlook Win32 Desktop Sends a NULL item whenever the item is changed, and will send the newly selected item afterwards. This is by design, and the NULL event must be handled. Is there a bug in the Typings that says this item cannot be NULL? **(Update)**: Our typescripts are indeed incorrect here. We will be updating the typings file to indicate that the item can be null. 

In another issues Shared Mailboxes are not supported at this time. *Delegate* shared folders Are supported. There is an issue where add-ins are showing up there, and we are evaluating how to fix this issue.

**Comment by mwickens** [comment]
> There is an issue where add-ins are showing up [for Shared Mailboxes], and we are evaluating how to fix this issue.

As mentioned, our product relies on the fact that this works (99.9% of the time) and the fact that Microsoft announced this as supported on UserVoice. I hope any fix is in the direction of official support, not removing functionality that lets our product work.

As for our original issue, as reported by @ArjunSureshKumar92, the problem has suddenly, inexplicably gone away for the one remaining customer who was experiencing it. We don't know of any change that accounts for this, unfortunately, so we're left wondering whether it might happen again.

**Comment by exextoc** [comment]
We will be updating our typings file to indicate that the item should be null. This was an error on our part. 

I don't have more to share about SharedMailboxes right now, but the product team is considering a number of different options, and your feedback has been noted.

**Comment by Emailgistics-Arjun** [comment]
Hi @ElizabethSamuel-MSFT ,
For a few months we did not face the issue and assumed the issue was resolved. However, we have one customer facing the same issue again. Multiple users are members of the shared mailbox, but only one user who is working from Europe (rest work in North America) is facing the issue. The error we get is the same as before. 

Any help will be much appreciated. Thanks in advance.
